File: server/ollama_app.py
from flask import Flask, request, jsonify
import requests

app = Flask(__name__)

@app.route('/', methods=['GET'])
def index():
    return jsonify("hello from ollama")

@app.route('/ask', methods=['POST'])
def ask():
    data = request.json
    prompt = data.get('prompt')
    print("prompt", prompt)
    if not prompt:
        return jsonify({"error": "No prompt provided"}), 400

    responses = []
    errors = []
    urls = [
        "http://localhost:11431/api/generate",
        "http://localhost:11434/api/generate",
        "http://localhost:11433/api/generate"
    ]
    models = [
        "llama3", 
        "phi3", 
        "mistral"
        ]

    for url, model in zip(urls, models):
        try:
            payload = {'model': model, 'prompt': prompt}
            headers = {"Content-Type": "application/json"}
            response = requests.post(url, json=payload, headers=headers)
            if response.status_code == 200:
                res_text = response.text.replace('}', '},').strip()
                if res_text.endswith(',,') or res_text.endswith('},'):
                    res_text = res_text[:-1]
                responses.append({
                    'model': model,
                    'resp':  '[' + res_text + ']'
                })
        except Exception as error:
            print(error)
            errors.append({'model': model, 'error': str(error)})

    return jsonify([responses, errors])

if __name__ == '__main__':
    app.run(port=5000)

File: server/ollama_app.js
const express = require('express');
const axios = require('axios');

const app = express();
app.use(express.json());

app.get('/', async (req, res) => {
    res.json("hello from ollama");
});

app.post('/ask', async (req, res) => {
    const data = req.body;
    const prompt = data.prompt;
    console.log(prompt);

    if (!prompt) {
        return res.status(400).json({ error: "No prompt provided" });
    }

    const responses = [];
    const errors = [];
    const urls = [
        "http://localhost:11431/api/generate",
        "http://localhost:11432/api/generate",
        "http://localhost:11433/api/generate",
    ];
    const models = [
        "llama3",
        "phi3",
        "mistral"
    ];

    for (let i = 0; i < urls.length; i++) {
        try {
            const url = urls[i];
            const payload = {
                model: models[i],
                prompt: prompt
            };
            const headers = {
                "Content-Type": "application/json"
            };
            const response = await axios.post(url, payload, { headers: headers });
            let resText = response.data;
            resText = resText.replace(/}/g, '},').trim();
            if (resText.endsWith(',,') || resText.endsWith('},')) {
                resText = resText.slice(0, -1);
            }
            if (response.status === 200) {
                responses.push({
                    model: models[i],
                    resp: JSON.parse('[' + resText + ']')
                });
            }
        } catch (error) {
            console.error(error);
            errors.push({
                model: models[i],
                error: error.message
            });
        }
        res.json([responses, errors]);
    }
});

app.listen(5000, () => {
    console.log('Server is running on port 5000');
});
